### PR TITLE
Replace directory with destination, allow specifying filename (#92)

### PR DESCRIPTION
**Pull request by ujjwalwahi**
@whimboo 
Replaced directory(-d, --directory) with destination(-d, --dest), now user can specify filename of the download content. Also updated test cases to test added functionality. This will fix issue #92.

### CODE REVIEW

File: mozdownload/scraper.py [hunk 1]
```diff
@@ -274,8 +279,14 @@ def total_seconds(td):
 
         attempt = 0
 
-        if not os.path.isdir(self.directory):
-            os.makedirs(self.directory)
+        directory_part = self.destination
+
+        # if file name is present in destination
+        if(self.target == self.destination):
```
Comment (whimboo) [inline mozdownload/scraper.py]: nit: Please remove the brackets. Those are not necessary here.

File: mozdownload/scraper.py [hunk 2]
```diff
@@ -274,8 +279,14 @@ def total_seconds(td):
...
+            directory_part = self.destination.rpartition('/')[0]
```
Comment (whimboo) [inline mozdownload/scraper.py]: There is `os.path.dirname()` which you can use here.

File: mozdownload/scraper.py [hunk 3]
```diff
@@ -274,8 +279,14 @@ def total_seconds(td):
 
         attempt = 0
 
-        if not os.path.isdir(self.directory):
-            os.makedirs(self.directory)
+        directory_part = self.destination
```
Comment (whimboo) [inline mozdownload/scraper.py]: I would put this into the else case to the if block below. Also please name the variable `directory` only.

File: mozdownload/scraper.py [hunk 4]
```diff
@@ -274,8 +279,14 @@ def total_seconds(td):
...
+
+        # if file name is present in destination
+        if(self.target == self.destination):
+            directory_part = self.destination.rpartition('/')[0]
+
+        if directory_part and not os.path.isdir(directory_part):
```
Comment (whimboo) [inline mozdownload/scraper.py]: Will there be a condition when we do not have a directory? I think we can omit this check.

File: mozdownload/scraper.py [hunk 5]
```diff
@@ -551,8 +562,14 @@ def __init__(self, url, *args, **kwargs):
     @property
     def target(self):
         target = urlparse(self.final_url)
-        filename = target.path.rpartition('/')[-1] or target.hostname
-        return os.path.join(self.directory, filename)
+        source_filename = target.path.rpartition('/')[-1] or target.hostname
+        target_file = os.path.join(self.destination, source_filename)
+
+        # if destination is path to file
+        source_file_ext = source_filename.rpartition('.')[-1]
```
Comment (whimboo) [inline mozdownload/scraper.py]: Please use `os.path.splitext()` to get a possible extension.

File: mozdownload/scraper.py [hunk 6]
```diff
@@ -551,8 +562,14 @@ def __init__(self, url, *args, **kwargs):
     @property
     def target(self):
         target = urlparse(self.final_url)
-        filename = target.path.rpartition('/')[-1] or target.hostname
-        return os.path.join(self.directory, filename)
+        source_filename = target.path.rpartition('/')[-1] or target.hostname
+        target_file = os.path.join(self.destination, source_filename)
```
Comment (whimboo) [inline mozdownload/scraper.py]: Similar to above. Please put this into an else block of the if condition below.

File: mozdownload/scraper.py [hunk 7]
```diff
@@ -887,11 +904,11 @@ def cli():
                       metavar='APPLICATION',
                       help='The name of the application to download, '
                            'default: "%default"')
-    parser.add_option('--directory', '-d',
-                      dest='directory',
+    parser.add_option('--dest', '-d',
```
Comment (whimboo) [inline mozdownload/scraper.py]: We should use the full-word for the `--` option.

File: tests/daily_scraper/test_daily_scraper.py [hunk 8]
```diff
@@ -256,6 +256,7 @@
     # -p mac64 --branch=mozilla-central
     {'args': {'application': 'b2g',
               'platform': 'win32',
+              'extension': 'zip',
```
Comment (whimboo) [inline tests/daily_scraper/test_daily_scraper.py]: We shouldn't have to change those tests, given that we fallback to the default extension in the scraper.

File: tests/daily_scraper/test_daily_scraper.py [hunk 9]
```diff
@@ -305,11 +316,20 @@ def test_scraper(self):
...
             self.assertEqual(scraper.target, expected_target)
+            destination_ext = DEFAULT_FILE_EXTENSIONS[entry['args']['platform']]
+            if 'extension' in entry['args']:
+                destination_ext = entry['args']['extension']
+            destination = os.path.join(self.temp_dir,"file." + destination_ext)
+            scraper2 = DailyScraper(destination=destination, base_url=self.wdir,
+                                   version=None, log_level='ERROR',
+                                   **entry['args'])
+            expected_target = destination
+            self.assertEqual(scraper2.target, expected_target)
```
Comment (whimboo) [inline tests/daily_scraper/test_daily_scraper.py]: Please move those additions out to its own test module, which is specifically about testing the `--destination` command line option.

File: tests/direct_scraper/test_direct_scraper.py [hunk 10]
```diff
@@ -40,6 +40,17 @@ def test_url_download(self):
...
+        scraper2 = DirectScraper(url=test_url,
+                                destination=os.path.join(self.temp_dir, 'file.html'),
+                                version=None,
+                                log_level='ERROR')
+        self.assertEqual(scraper2.target,
+                         os.path.join(self.temp_dir, 'file.html'))
+
+        scraper2.download()
+        self.assertTrue(os.path.isfile(os.path.join(self.temp_dir,
+                                                    scraper2.target)))
```
Comment (whimboo) [inline tests/direct_scraper/test_direct_scraper.py]: It will be enough to add a test for the base scraper in case of a filename as destination because the logic is contained only there.

File: tests/release_candidate_scraper/test_release_candidate_scraper.py [hunk 11]
```diff
@@ -130,12 +137,22 @@ def test_scraper(self):
...
+            destination_ext = DEFAULT_FILE_EXTENSIONS[entry['args']['platform']]
+            if 'extension' in entry['args']:
+                destination_ext = entry['args']['extension']
+            destination = os.path.join(self.temp_dir,"file." + destination_ext)
+            scraper2 = ReleaseCandidateScraper(destination=destination,
+                                              base_url=self.wdir,
+                                              log_level='ERROR',
+                                              **entry['args'])
+            expected_target = destination
+            self.assertEqual(scraper2.target, expected_target)
```
Comment (whimboo) [inline tests/release_candidate_scraper/test_release_candidate_scraper.py]: Same here as above.

File: tests/release_scraper/test_release_scraper.py [hunk 12]
```diff
@@ -128,10 +134,18 @@ def test_scraper(self):
...
             expected_target = os.path.join(self.temp_dir, entry['target'])
             self.assertEqual(scraper.target, expected_target)
+            destination_ext = DEFAULT_FILE_EXTENSIONS[entry['args']['platform']]
+            if 'extension' in entry['args']:
+                destination_ext = entry['args']['extension']
+            destination = os.path.join(self.temp_dir,"file." + destination_ext)
+            scraper2 = ReleaseScraper(destination=destination, base_url=self.wdir,
+                                     log_level='ERROR', **entry['args'])
+            expected_target = destination
+            self.assertEqual(scraper2.target, expected_target)
```
Comment (whimboo) [inline tests/release_scraper/test_release_scraper.py]: Same as above.

File: tests/tinderbox_scraper/test_tinderbox_scraper.py [hunk 13]
```diff
@@ -240,11 +247,20 @@ def test_scraper(self):
...
             self.assertEqual(scraper.target, expected_target)
+            destination_ext = DEFAULT_FILE_EXTENSIONS[entry['args']['platform']]
+            if 'extension' in entry['args']:
+                destination_ext = entry['args']['extension']
+            destination = os.path.join(self.temp_dir, "file." + destination_ext)
+            scraper2 = TinderboxScraper(destination=destination, version=None,
+                                       base_url=self.wdir, log_level='ERROR',
+                                       **entry['args'])
+            expected_target = destination
+            self.assertEqual(scraper2.target, expected_target)
```
Comment (whimboo) [inline tests/tinderbox_scraper/test_tinderbox_scraper.py]: Same as above.

**Comment by whimboo** [comment]
Good start @ujjwalwahi! The changes in the base code looks fine and need some smaller updates. For the tests a bit more work will still be necessary. Therefore please see my inline comments. If you have questions don't hesitate to ask me here or on IRC. Thanks!

**Comment by whimboo** [comment]
Hi @ujjwalwahi. I hope all is fine for you? I would like to know if you can find the time to address my review comments. Or if you have remaining questions. Thanks

**Comment by ujjwalwahi** [comment]
Hi @whimboo . I am working on this, little busy with my exams. I updated scrapper.py as per reviews, no issue there. Working with tests. Will deliver it soon.

**Comment by whimboo** [comment]
Good to hear that! But given that you are in exams please concentrate on those for now. This PR can wait.

**Comment by ujjwalwahi** [comment]
@whimboo 
Made changes according to review

File: mozdownload/scraper.py [hunk 14]
```diff
@@ -274,15 +279,16 @@ def total_seconds(td):
...
-            os.makedirs(self.directory)
-
         # Don't re-download the file
         if os.path.isfile(os.path.abspath(self.target)):
             self.logger.info("File has already been downloaded: %s" %
                              (self.target))
             return
 
+        directory = os.path.dirname(self.target)
+        if directory and not os.path.isdir(directory):
```
Comment (whimboo) [inline mozdownload/scraper.py]: We always have a valid directory here. It should never be `None`. So you can get rid of the first check.

File: mozdownload/scraper.py [hunk 15]
```diff
@@ -97,7 +97,7 @@ def __init__(self, directory, version, platform=None,
         self._target = None
         self._binary = None
 
-        self.directory = directory
+        self.destination = destination
```
Comment (whimboo) [inline mozdownload/scraper.py]: I think with this rename we have a duplication between `self.destination` and `self.target`. I assume we should get rid of `self.target`. Maybe we can make `self.destination` a write property, which contains the code as currently lives in `self.target`. I don't see why we would have to run the code more than once for the same session, especially when no input parameters are changing.

File: mozdownload/scraper.py [hunk 16]
```diff
@@ -887,11 +900,11 @@ def cli():
...
                            'default: "%default"')
-    parser.add_option('--directory', '-d',
-                      dest='directory',
+    parser.add_option('--destination', '-d',
+                      dest='destination',
                       default=os.getcwd(),
-                      metavar='DIRECTORY',
-                      help='Target directory for the download, default: '
+                      metavar='DESTINATION',
+                      help='Target destination for the download, default: '
```
Comment (whimboo) [inline mozdownload/scraper.py]: Maybe `Directory or file name to download the file to`?

File: tests/base_scraper/test_base_scraper.py [hunk 17]
```diff
@@ -66,7 +66,15 @@ def test_download(self):
...
-        scraper = mozdownload.Scraper(directory=self.temp_dir,
+        filename = 'download_test.txt'
+        # standard download
+        test_url = urljoin(self.wdir, 'download_test.txt')
+        scraper = mozdownload.DirectScraper(url=test_url,
+                                            destination=self.temp_dir,
+                                            version=None,
+                                            log_level='ERROR')
+        scraper.download()
+        scraper = mozdownload.Scraper(destination=self.temp_dir,
```
Comment (whimboo) [inline tests/base_scraper/test_base_scraper.py]: I asked about those changes in my last review. Maybe you have missed to remove them? This is not related to the test method you placed the code into.

File: tests/base_scraper/test_base_scraper.py [hunk 18]
```diff
@@ -74,5 +82,28 @@ def test_notimplementedexceptions(self):
         self.assertRaises(mozdownload.NotImplementedError,
                           scraper.build_filename, 'invalid binary')
 
+    def test_destination(self):
+        """Test for various destination scenarios"""
+
+        filename = 'download_test.txt'
+        test_url = urljoin(self.wdir, 'download_test.txt')
+
+        #destination is directory
```
Comment (whimboo) [inline tests/base_scraper/test_base_scraper.py]: nit: please add a blank after the #.

File: tests/base_scraper/test_base_scraper.py [hunk 19]
```diff
@@ -74,5 +82,28 @@ def test_notimplementedexceptions(self):
         self.assertRaises(mozdownload.NotImplementedError,
                           scraper.build_filename, 'invalid binary')
 
+    def test_destination(self):
+        """Test for various destination scenarios"""
+
+        filename = 'download_test.txt'
+        test_url = urljoin(self.wdir, 'download_test.txt')
```
Comment (whimboo) [inline tests/base_scraper/test_base_scraper.py]: Please use the formerly defined filename to join here.

File: tests/base_scraper/test_base_scraper.py [hunk 20]
```diff
@@ -74,5 +82,28 @@ def test_notimplementedexceptions(self):
...
+
+        #destination is directory
+        scraper = mozdownload.DirectScraper(url=test_url,
+                                            destination=self.temp_dir,
+                                            version=None,
+                                            log_level='ERROR')
+        expected_target = os.path.join(self.temp_dir, filename)
+        self.assertEqual(scraper.target, expected_target)
+
+        #destination is file
```
Comment (whimboo) [inline tests/base_scraper/test_base_scraper.py]: Same here as above.

File: tests/base_scraper/test_base_scraper.py [hunk 21]
```diff
@@ -74,5 +82,28 @@ def test_notimplementedexceptions(self):
...
+        filename = 'download_test.txt'
+        test_url = urljoin(self.wdir, 'download_test.txt')
+
+        #destination is directory
+        scraper = mozdownload.DirectScraper(url=test_url,
+                                            destination=self.temp_dir,
+                                            version=None,
+                                            log_level='ERROR')
+        expected_target = os.path.join(self.temp_dir, filename)
+        self.assertEqual(scraper.target, expected_target)
```
Comment (whimboo) [inline tests/base_scraper/test_base_scraper.py]: There is no need to define `expected_target` first. Just put it directly inside the call to `assertEqual`.

File: tests/base_scraper/test_base_scraper.py [hunk 22]
```diff
@@ -74,5 +82,28 @@ def test_notimplementedexceptions(self):
...
 
+    def test_destination(self):
+        """Test for various destination scenarios"""
+
+        filename = 'download_test.txt'
+        test_url = urljoin(self.wdir, 'download_test.txt')
+
+        #destination is directory
+        scraper = mozdownload.DirectScraper(url=test_url,
+                                            destination=self.temp_dir,
```
Comment (whimboo) [inline tests/base_scraper/test_base_scraper.py]: Please add another test which has a non-existent folder as destination, so we can test that the path is correctly created.

File: tests/base_scraper/test_base_scraper.py [hunk 23]
```diff
@@ -74,5 +82,28 @@ def test_notimplementedexceptions(self):
...
+        expected_target = os.path.join(self.temp_dir, filename)
+        self.assertEqual(scraper.target, expected_target)
+
+        #destination is file
+        destination = os.path.join(self.temp_dir, filename)
+        scraper = mozdownload.DirectScraper(url=test_url,
+                                            destination=destination,
+                                            version=None,
+                                            log_level='ERROR')
+        expected_target = destination
```
Comment (whimboo) [inline tests/base_scraper/test_base_scraper.py]: There is no need for this line. Use `destination` directly in the `assertEqual` call.

File: tests/daily_scraper/test_daily_scraper.py (hunk 8 above)
Comment (whimboo) [inline tests/daily_scraper/test_daily_scraper.py]: This still applies and need to be removed.

File: tests/daily_scraper/test_daily_scraper.py [hunk 24]
```diff
@@ -263,6 +264,7 @@
     # -p win32 --branch=mozilla-central -l en-US
     {'args': {'application': 'b2g',
               'platform': 'win32',
+              'extension': 'zip',
```
Comment (whimboo) [inline tests/daily_scraper/test_daily_scraper.py]: Same for this and all the others here.

File: tests/daily_scraper/test_daily_scraper.py [hunk 25]
```diff
@@ -305,14 +316,36 @@ def test_scraper(self):
...
-            scraper = DailyScraper(directory=self.temp_dir, base_url=self.wdir,
+            scraper = DailyScraper(destination=self.temp_dir, base_url=self.wdir,
                                    version=None, log_level='ERROR',
                                    **entry['args'])
-            expected_target = os.path.join(self.temp_dir, entry['target'])
-            self.assertEqual(scraper.target, expected_target)
             self.assertEqual(urllib.unquote(scraper.final_url),
                 urljoin(self.wdir, entry['target_url']))
 
+    def test_destination(self):
```
Comment (whimboo) [inline tests/daily_scraper/test_daily_scraper.py]: There is no need to test this here. `target` and `destination` are properties of the `BaseScraper` class and should only be tested there.

File: tests/release_candidate_scraper/test_release_candidate_scraper.py [hunk 26]
```diff
@@ -122,6 +122,13 @@
 
 tests = firefox_tests + thunderbird_tests
 
+DEFAULT_FILE_EXTENSIONS = {'linux': 'tar.bz2',
+                           'linux64': 'tar.bz2',
+                           'mac': 'dmg',
+                           'mac64': 'dmg',
+                           'win32': 'exe',
+                           'win64': 'exe'}
```
Comment (whimboo) [inline tests/release_candidate_scraper/test_release_candidate_scraper.py]: Please remove those lines.

File: tests/release_scraper/test_release_scraper.py [hunk 27]
```diff
@@ -120,6 +120,12 @@
 
 tests = firefox_tests + thunderbird_tests
 
+DEFAULT_FILE_EXTENSIONS = {'linux': 'tar.bz2',
+                           'linux64': 'tar.bz2',
+                           'mac': 'dmg',
+                           'mac64': 'dmg',
+                           'win32': 'exe',
+                           'win64': 'exe'}
```
Comment (whimboo) [inline tests/release_scraper/test_release_scraper.py]: Same for those.

File: tests/tinderbox_scraper/test_tinderbox_scraper.py [hunk 28]
```diff
@@ -232,6 +232,13 @@
 
 tests = firefox_tests + thunderbird_tests
 
+DEFAULT_FILE_EXTENSIONS = {'linux': 'tar.bz2',
+                           'linux64': 'tar.bz2',
+                           'mac': 'dmg',
+                           'mac64': 'dmg',
+                           'win32': 'exe',
+                           'win64': 'exe'}
+
```
Comment (whimboo) [inline tests/tinderbox_scraper/test_tinderbox_scraper.py]: Same here.

**Comment by ujjwalwahi** [comment]
@whimboo 
Worked on reviews.

File: tests/base_scraper/test_base_scraper.py [hunk 29]
```diff
@@ -113,5 +113,35 @@ def test_optional_authentication(self):
...
+
+        # destination is file
+        destination = os.path.join(self.temp_dir, filename)
+        scraper = mozdownload.DirectScraper(url=test_url,
+                                            destination=destination,
+                                            version=None,
+                                            log_level='ERROR')
+        self.assertEqual(scraper.target, destination)
+
+        # destination directory does not exist
```
Comment (whimboo) [inline tests/base_scraper/test_base_scraper.py]: I would also like to see the case when a filename has been specified and the folder does not exist yet. Maybe even multiple levels deep, so we can ensure that we correctly create a folder tree.

**Comment by whimboo** [comment]
Except the one missing test this PR looks great! I also tested it and weren't able to find a problem. @davehunt would you mind to have a look at, so we can make sure this change satisfies your needs?

**Comment by ujjwalwahi** [comment]
@whimboo 
Added the required test section

**Comment by whimboo** [comment]
This looks great now! Thanks for your tireless work @ujjwalwahi! Given that Travis is also happy this time let me get it merged in a bit.

**Comment by whimboo** [comment]
When squashing the branch I got some rebase problems, so I had to fix those before pushing out. Finally the PR got merged as https://github.com/mozilla/mozdownload/commit/942d9728ec2eafd58ad234c1d66d2ff46de2d86e.
